lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,557 @@
+	}
+
+	// Disconnect the internal transactions in reverse.
+	for ii := len(txMeta.Txns) - 1; ii >= 0; ii-- {
+		innerTxn := txMeta.Txns[ii]
+
+		if err := bav.DisconnectTransaction(
+			innerTxn,
+			innerTxn.Hash(),
+			operationData.AtomicTxnsInnerUtxoOps[ii],
+			blockHeight); err != nil {
+			return errors.Wrapf(err, "_disconnectAtomicTxnsWrapper")
+		}
+	}
+
+	return nil
+}
+
+//
+// TYPES: AtomicTxnsWrapperTxindexMetadata
+//
+
+type AtomicTxnsWrapperTxindexMetadata struct {
+	InnerTxnsTransactionMetadata []*TransactionMetadata
+}
+
+func (txindexMetadata *AtomicTxnsWrapperTxindexMetadata) RawEncodeWithoutMetadata(
+	blockHeight uint64,
+	skipMetadata ...bool,
+) []byte {
+	var data []byte
+	data = append(data, UintToBuf(uint64(len(txindexMetadata.InnerTxnsTransactionMetadata)))...)
+	for _, innerMetadata := range txindexMetadata.InnerTxnsTransactionMetadata {
+		txnBytes := innerMetadata.RawEncodeWithoutMetadata(blockHeight, skipMetadata...)
+		data = append(data, UintToBuf(uint64(len(txnBytes)))...)
+		data = append(data, txnBytes...)
+	}
+	return data
+}
+
+func (txindexMetadata *AtomicTxnsWrapperTxindexMetadata) RawDecodeWithoutMetadata(
+	blockHeight uint64,
+	rr *bytes.Reader,
+) error {
+	// Read the number of inner transactions.
+	numTxns, err := ReadUvarint(rr)
+	if err != nil {
+		return errors.Wrap(err,
+			"AtomicTxnsWrapperTxindexMetadata.RawDecodeWithoutMetadata: Problem reading numTxns")
+	}
+	txindexMetadata.InnerTxnsTransactionMetadata, err = SafeMakeSliceWithLength[*TransactionMetadata](numTxns)
+	if err != nil {
+		return errors.Wrap(err,
+			"AtomicTxnsWrapperTxindexMetadata.RawDecodeWithoutMetadata: "+
+				"Problem allocating InnerTxnsTransactionMetadata")
+	}
+
+	// Read the transactions.
+	for ii := uint64(0); ii < numTxns; ii++ {
+		txindexMetadata.InnerTxnsTransactionMetadata[ii] = &TransactionMetadata{}
+
+		// Figure out how many bytes are associated with the ith transaction metadata.
+		numTxnMetadataBytes, err := ReadUvarint(rr)
+		if err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperTxindexMetadata.RawDecodeWithoutMetadata: "+
+					"Problem reading number of bytes in transaction metadata")
+		}
+
+		// Allocate memory for the transaction metadata bytes to be read into.
+		txnMetadataBytes, err := SafeMakeSliceWithLength[byte](numTxnMetadataBytes)
+		if err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperTxindexMetadata.RawDecodeWithoutMetadata: "+
+					"Problem allocating bytes for transaction")
+		}
+
+		// Read the transaction metadata into the txnBytes memory buffer.
+		if _, err = io.ReadFull(rr, txnMetadataBytes); err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperTxindexMetadata.RawDecodeWithoutMetadata: Problem reading bytes for transaction")
+		}
+
+		// Convert the txnBytes buffer to a TransactionMetadata struct.
+		if err = txindexMetadata.InnerTxnsTransactionMetadata[ii].RawDecodeWithoutMetadata(blockHeight, rr); err != nil {
```

Comment:
This looks wrong. You're reading past the bytes you want on line 537. I'll leave my suggested changes in another comment.